wordpress-mobile/WordPress-iOS
# [Auto-upload Published] Cancelation Flow

Part of #12240. In #12252, we added auto-uploading of locally published posts. In #12278, we updated the Post List and Notice messages for it. This PR adds the cancelation flow. 

![image](https://user-images.githubusercontent.com/198826/63067091-8b39c100-beca-11e9-9d62-cdbcb52a7db0.png)

Credit goes to @diegoreymendez. Most of this was taken from his work on #12178.  

## What Cancelation Means

In the context automatic uploads, canceling means:

- Leave the changes as local changes
- Do not automatically upload

Canceling should not discard the local changes. In WPAndroid, cancelation is implemented as [turning off the _confirmation flag_](https://github.com/wordpress-mobile/WordPress-Android/blob/1bdae360a9c6213eea84d6bbfc71c3c582a630c6/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadUtils.java#L442-L449). 

## UploadActionUseCase

In this PR, I introduced a new class, `UploadActionUseCase`. Its goal is to encapsulate the business logic of:

- Should a post be auto-uploaded or just remote auto-saved? 
- Can the user cancel the auto-upload? 

It is shared between `PostCoordinator` and Post List classes.

![image](https://user-images.githubusercontent.com/198826/63098587-7132c980-bf30-11e9-9da6-5a46c850ae65.png)

I also envision this to be the location of the safety features that we need to add. And perhaps a feature flag to disable auto-uploading. 

## What's Missing

There are still a few parts missing here. I wanted to focus on the main flow of confirmation → upload and confirmation → cancel. Here are some that were not added in this PR:

- Notice should show Cancel button #12319 
- Show notice after canceling #12325 

And for completing #12240, we still need to implement these:

- Safety: Content hash and timeout for offline actions #12136 
- Show Publish button in Post List #12321
- Automatically upload published and draft posts with existing remote #12324 
- Remote auto-save unconfirmed draft and published posts #12322 
- Fix the “Upload failed“ shown after canceling. 

Some are in the checklists of #12240. 

### Stub for Confirmation Property

Jan is working on #12136 to adding a content hash property that will be used to determine whether the user confirmed or canceled the auto-upload. In this PR, I added a stub:

https://github.com/wordpress-mobile/WordPress-iOS/blob/504b9153cc2ed2ac29e9468b746d64919bfc74cc/WordPress/Classes/Services/PostCoordinator%2BUploadActionUseCase.swift#L82

I riddled the code with `#warning()` so I won't forget to remove or replace them. The stub works by adding or removing a `"<C>"` to the `postTitle`. For example, if the user entered `"Llamas"` as the title:

Action | Saved `postTitle`
--------|-------
Confirmed auto-upload        |       `<C> Llamas`
Canceled | `LLamas`

@jklausa I'd like to get your input on how this will integrate with what you've implemented so far for #12136. I'll help you out in integrating the work. 

## Testing

1. Go offline.
2. Create a post and publish it. 
3. Press the _Cancel_ button in the Post List. 

The post should not be auto-uploaded. 

## Reviewing 

Since this is a critical part of the publishing flow, it would be great if 2 developers can review it. Please let me know if you think a sync call explaining what's been done so far would be beneficial. We can arrange that. 

## Release Notes
